Missing button in Source Control panel to initialize Fossil repo

First of all, thanks a lot for this extension!

On my system (openSUSE Tumbleweed, Visual Studio Code 1.67.1) I don’t seem to have a button to initialize a Fossil repository directly from the Source Control panel (expected place in the red square):
![Screenshot_20220517_184915](https://user-images.githubusercontent.com/2722396/168867059-09c5288b-2b78-49b2-b1aa-52745f4bf353.png)
contrary to what is explained in the README, with the Mercurial gif:
![README](https://raw.githubusercontent.com/koog1000/vscode-fossil/master/images/init.gif)

I did some further testing, and:

* The issue is the same whether I have a folder opened in VSCode or not
* In order to get the extension to get going, I have to manually init then open a Fossil repository through the command line, and finally create a file in the folder: the extension will detect that there is a new untracked file and a Fossil repository, and the Source Control panel will work as expected from this point on
* Installing Mercurial on my system and mrcrowl’s Hg extension correctly shows the Mercurial button, so it seems to be an issue specific to the forked Fossil extension (maybe changes in VSCode source control panel that the Hg extension has adapted to, but not yet this one?) 
![Screenshot_20220517_182257](https://user-images.githubusercontent.com/2722396/168867182-34de551b-547a-49dd-842a-686e35bb3642.png)

Hi @senyai, thanks a lot for the fixes, I tried building your PR in order to test it but it seems I have an issue with the npm build process on my system (unfortunately I am not really knowledgeable about Node.js / npm / TypeScript): after getting the source code, I tried building with `npm install` as shown in the README but I get the following error

```
> fossil@0.1.2 postinstall
> node ./node_modules/vscode/bin/install

node:internal/modules/cjs/loader:936
  throw err;
  ^

Error: Cannot find module '/home/xxx/.vscode/extensions/vscode-fossil-restore-init/node_modules/vscode/bin/install'
    at Function.Module._resolveFilename (node:internal/modules/cjs/loader:933:15)
    at Function.Module._load (node:internal/modules/cjs/loader:778:27)
    at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:77:12)
    at node:internal/main/run_main_module:17:47 {
  code: 'MODULE_NOT_FOUND',
  requireStack: []
}
```

And indeed, I don’t have that expected "vscode" folder inside "node_modules". Do you have an idea of what the problem could be? Is there something I am missing? I installed Node.js 16.14.2 and npm 8.5.0. I also tried with Node.js 17.7.1 and npm 8.5.2 (the other versions available in openSUSE Tumbleweed), but I had the same error.

--- a/.gitignore
+++ b/.gitignore
@@ -7,3 +7,4 @@ test/sandbox
 resources/icons/*-dark
 resources/icons/*-light
 *.js
+package-lock.json

--- a/.gitignore
+++ b/.gitignore
@@ -6,3 +6,4 @@ images/production
 test/sandbox
 resources/icons/*-dark
 resources/icons/*-light
+*.js

--- a/.vscode/launch.json
+++ b/.vscode/launch.json
@@ -7,7 +7,9 @@
             "type": "extensionHost",
             "request": "launch",
             "runtimeExecutable": "${execPath}",
-            "args": ["--extensionDevelopmentPath=${workspaceRoot}" ],
+            "args": [
+                "--extensionDevelopmentPath=${workspaceRoot}"
+            ],
             "stopOnEntry": false,
             "sourceMaps": true,
             "outFiles": [
@@ -24,7 +26,9 @@
                 "--extensionDevelopmentPath=${workspaceFolder}",
                 "--extensionTestsPath=${workspaceFolder}/out/test/suite/index"
             ],
-            "outFiles": ["${workspaceFolder}/out/test/**/*.js"]
+            "outFiles": [
+                "${workspaceFolder}/out/test/**/*.js"
+            ]
         }
     ]
 }

--- a/.vscode/tasks.json
+++ b/.vscode/tasks.json
@@ -27,7 +27,7 @@
     },
 
     // we run the custom script "compile" as defined in package.json
-    "args": ["run", "compile", "--loglevel", "debug"],
+    "args": ["run", "compile", "--loglevel", "info"],
 
     // The tsc compiler is started in watching mode
     "isBackground": true,

--- a/.vscodeignore
+++ b/.vscodeignore
@@ -1,2 +1,5 @@
 .vscode
 **/*.ts
+**/.gitignore
+**/*.map
+images/

--- a/README.md
+++ b/README.md
@@ -125,10 +125,10 @@ You will need to install [Node.js](https://nodejs.org/en/download/)
 on your computer and add it to your `$PATH`.
 
 ### Build Steps
-1. `git clone` repository and place it in your VSCode extension folder (~/.vscode/extensions/ or similar).
-2. `npm install` from clone directory to install local dependencies
+1. `git clone` repository anywhere on your filesystem.
+2. `npm install` from clone directory to install local dependencies.
 3. `npm run compile` to build extension.
-
+4. press `F5` to run the extension.
 
 # Acknowledgements
 

--- a/package.json
+++ b/package.json
@@ -16,7 +16,7 @@
     "bugs": {
         "url": "https://github.com/koog1000/vscode-fossil/issues"
     },
-    "version": "0.1.1",
+    "version": "0.1.2",
     "engines": {
         "vscode": "^1.36.0"
     },
@@ -27,7 +27,7 @@
         "scm"
     ],
     "icon": "icon.png",
-    "enableProposedApi": false,
+    "enabledApiProposals": [],
     "categories": [
         "SCM Providers",
         "Other"
@@ -130,6 +130,15 @@
             {
                 "command": "fossil.init",
                 "title": "%command.init%",
+                "category": "Fossil",
+                "icon": {
+                    "light": "icon.png",
+                    "dark": "icon.png"
+                }
+            },
+            {
+                "command": "fossil.open",
+                "title": "%command.open%",
                 "category": "Fossil",
                 "icon": {
                     "light": "icon.png",
@@ -460,7 +469,17 @@
                 {
                     "command": "fossil.init",
                     "group": "navigation",
-                    "when": "config.fossil.enabled && scmProvider == '' && fossilOpenRepositoryCount == 0"
+                    "when": "config.fossil.enabled && !scmProvider && fossilOpenRepositoryCount == 0"
+                },
+                {
+                    "command": "fossil.open",
+                    "group": "0_open",
+                    "when": "config.fossil.enabled && !scmProvider && fossilOpenRepositoryCount == 0"
+                },
+                {
+                    "command": "fossil.clone",
+                    "when": "config.fossil.enabled",
+                    "group": "0_open"
                 },
                 {
                     "command": "fossil.commit",
@@ -798,15 +817,15 @@
         "postinstall": "node ./node_modules/vscode/bin/install"
     },
     "dependencies": {
-        "@types/mocha": "^5.2.7",
         "glob": "^7.2.0",
-        "tmp": "0.0.31",
-        "vscode-nls": "^2.0.1",
-        "vscode-test": "^1.6.1"
+        "tmp": "0.2.1",
+        "vscode-nls": "^5.0.1"
     },
     "devDependencies": {
-        "@types/node": "^7.0.4",
-        "@types/tmp": "0.0.33",
+        "vscode-test": "^1.6.1",
+        "@types/mocha": "^9.1.0",
+        "@types/node": "^10.17.24",
+        "@types/tmp": "^0.2.1",
         "@types/vscode": "^1.36.0"
     },
     "__metadata": {

--- a/package.json
+++ b/package.json
@@ -812,9 +812,10 @@
         }
     },
     "scripts": {
-        "vscode:prepublish": "tsc -p ./",
+        "vscode:prepublish": "npm run compile",
         "compile": "tsc -watch -p ./",
-        "postinstall": "node ./node_modules/vscode/bin/install"
+        "package": "vsce package",
+        "deploy": "vsce publish"
     },
     "dependencies": {
         "glob": "^7.2.0",

--- a/package.nls.json
+++ b/package.nls.json
@@ -1,6 +1,7 @@
 {
-    "command.clone": "Clone",
-    "command.init": "Initialize Repository (Fossil)",
+    "command.clone": "Clone Fossil Repository",
+    "command.init": "Initialize Fossil Repository",
+    "command.open": "Open Fossil Repository",
     "command.close": "Close Repository",
     "command.openui": "Open web UI",
     "command.refresh": "Refresh",

--- a/src/commands.ts
+++ b/src/commands.ts
@@ -10,7 +10,7 @@ import {Uri, commands, scm, Disposable, window, workspace, OutputChannel,
 import { LineChange, revertChanges } from './revert';
 import * as nls from "vscode-nls";
 import * as path from "path";
-import { Ref, Fossil, Commit, FossilError, FossilErrorCodes, IFileStatus, CommitDetails } from "./fossilBase";
+import { Ref, Fossil, Commit, FossilError, FossilErrorCodes, IFileStatus, CommitDetails, FossilPath,FossilRoot, FossilURI } from "./fossilBase";
 import { Model } from "./model";
 import { Resource, Status, CommitOptions, CommitScope, MergeStatus, LogEntriesOptions, Repository } from "./repository"
 import { WorkingDirectoryGroup, StagingGroup, MergeGroup, UntrackedGroup, isResourceGroup } from "./resourceGroups";
@@ -197,7 +197,7 @@ export class CommandCenter {
             const regex = (url.search('@') < 0) ? /(^.+\:\/\/)(.+)/ : /(^.+\:\/\/).*@(.+)/
             const match = url.match(regex);
             if(match){
-                url = match[1] + username + ':' + userauth + '@' + match[2]
+                url = match[1] + username + ':' + userauth + '@' + match[2] as FossilURI
             }
         }
         const parentPath = await interaction.inputCloneParentPath();
@@ -208,37 +208,61 @@ export class CommandCenter {
         const clonePromise = this.fossil.clone(url, parentPath);
         interaction.statusCloning(clonePromise);
 
+        const repositoryPath = await clonePromise;
+        await this.askOpenRepository(repositoryPath, parentPath);
+    }
+
+    async openRepository(filePath: FossilPath, parentPath: FossilRoot) : Promise<void> {
         try {
-            const repositoryPath = await clonePromise;
-            const openClonedRepo = await interaction.promptOpenClonedRepo();
-            if (openClonedRepo) {
-                this.fossil.openClone(repositoryPath, parentPath)
+            await this.fossil.openClone(filePath, parentPath);
+        } catch (err) {
+            if (err instanceof FossilError && err.fossilErrorCode === FossilErrorCodes.OperationMustBeforced) {
+                const openNotEmpty = await interaction.confirmOpenNotEmpty(parentPath);
+                if (openNotEmpty) {
+                    this.fossil.openCloneForce(filePath, parentPath);
+                }
+            } else {
+                throw err;
             }
         }
-        catch (err) {
-            throw err;
+    }
+
+    /**
+     * ask user to open the repo after `clone` or `init`
+     */
+    async askOpenRepository(filePath: FossilPath, fossilRoot: FossilRoot) : Promise<void> {
+        const openClonedRepo = await interaction.promptOpenClonedRepo();
+        if (openClonedRepo) {
+            await this.openRepository(filePath, fossilRoot);
         }
     }
 
     @command('fossil.init')
     async init(): Promise<void> {
-        const result = await interaction.openFileDialog(false, true, false, localize('init repo', 'Respository Folder'))
+        const fossilFilePath = await interaction.initFileDialog()
 
-        if (!result || result.length === 0) {
+        if (fossilFilePath === undefined) {
             return;
         }
+        const dirname = path.dirname(fossilFilePath) as FossilRoot;
 
-        const uri = result[0];
-        const path = uri.fsPath;
-        const fileName = await interaction.inputRepoName()
-        if(fileName && fileName.length){
-            await this.fossil.init(path, fileName);
-            const openClonedRepo = await interaction.promptOpenClonedRepo();
-            if (openClonedRepo) {
-                this.fossil.openClone(fileName, path)
-            }
+        // run init in the file folder in case any any artifacts appear
+        await this.fossil.init(dirname, fossilFilePath);
+        await this.askOpenRepository(fossilFilePath, dirname);
+        await this.model.tryOpenRepository(dirname);
+    }
+
+    @command('fossil.open')
+    async open(): Promise<void> {
+        const uri = await interaction.openFileDialog()
+        if (uri === undefined) {
+            return;
         }
-        await this.model.tryOpenRepository(path);
+        const parentPath = await interaction.inputCloneParentPath();
+        if (!parentPath) {
+            return;
+        }
+        await this.openRepository(uri, parentPath);
     }
 
     @command('fossil.close', { repository: true})
@@ -989,6 +1013,7 @@ export class CommandCenter {
                 || repository.mergeGroup.getResource(uri)
                 || repository.conflictGroup.getResource(uri);
         }
+        return undefined;
     }
 
     // private runByRepository<T>(resource: Uri, fn: (repository: Repository, resource: Uri) => Promise<T>): Promise<T[]>;

--- a/src/commands.ts
+++ b/src/commands.ts
@@ -200,15 +200,14 @@ export class CommandCenter {
                 url = match[1] + username + ':' + userauth + '@' + match[2] as FossilURI
             }
         }
-        const fossilRoot = await interaction.selectFossilRootPath();
-        if (!fossilRoot) {
+        const fossilPath = await interaction.selectNewFossilPath('Clone');
+        if (!fossilPath) {
             return;
         }
 
-        const clonePromise = this.fossil.clone(url, fossilRoot);
+        const clonePromise = this.fossil.clone(url, fossilPath);
         interaction.statusCloning(clonePromise);
-
-        const fossilPath = await clonePromise;
+        const fossilRoot = await clonePromise;
         await this.askOpenRepository(fossilPath, fossilRoot);
     }
 
@@ -244,7 +243,7 @@ export class CommandCenter {
 
     @command('fossil.init')
     async init(): Promise<void> {
-        const fossilPath = await interaction.selectNewFossilPath();
+        const fossilPath = await interaction.selectNewFossilPath('Create');
 
         if (!fossilPath) {
             return;

--- a/src/commands.ts
+++ b/src/commands.ts
@@ -200,7 +200,7 @@ export class CommandCenter {
                 url = match[1] + username + ':' + userauth + '@' + match[2] as FossilURI
             }
         }
-        const parentPath = await interaction.inputCloneParentPath();
+        const parentPath = await interaction.inputFossilRootPath();
         if (!parentPath) {
             return;
         }
@@ -212,6 +212,10 @@ export class CommandCenter {
         await this.askOpenRepository(repositoryPath, parentPath);
     }
 
+    /**
+     * Execute "fossil open". When FossilRoot has files allow to
+     * run "fossil open --force"
+     */
     async openRepository(filePath: FossilPath, parentPath: FossilRoot) : Promise<void> {
         try {
             await this.fossil.openClone(filePath, parentPath);
@@ -228,7 +232,7 @@ export class CommandCenter {
     }
 
     /**
-     * ask user to open the repo after `clone` or `init`
+     * ask user to run "fossil open" after `clone` or `init`
      */
     async askOpenRepository(filePath: FossilPath, fossilRoot: FossilRoot) : Promise<void> {
         const openClonedRepo = await interaction.promptOpenClonedRepo();
@@ -244,25 +248,26 @@ export class CommandCenter {
         if (fossilFilePath === undefined) {
             return;
         }
-        const dirname = path.dirname(fossilFilePath) as FossilRoot;
+        const rootPath = path.dirname(fossilFilePath) as FossilRoot;
 
         // run init in the file folder in case any any artifacts appear
-        await this.fossil.init(dirname, fossilFilePath);
-        await this.askOpenRepository(fossilFilePath, dirname);
-        await this.model.tryOpenRepository(dirname);
+        await this.fossil.init(rootPath, fossilFilePath);
+        await this.askOpenRepository(fossilFilePath, rootPath);
+        await this.model.tryOpenRepository(rootPath);
     }
 
     @command('fossil.open')
     async open(): Promise<void> {
-        const uri = await interaction.openFileDialog()
-        if (uri === undefined) {
+        const fossilPath = await interaction.openFileDialog()
+        if (fossilPath === undefined) {
             return;
         }
-        const parentPath = await interaction.inputCloneParentPath();
-        if (!parentPath) {
+        const rootPath = await interaction.inputFossilRootPath();
+        if (!rootPath) {
             return;
         }
-        await this.openRepository(uri, parentPath);
+        await this.openRepository(fossilPath, rootPath);
+        await this.model.tryOpenRepository(rootPath);
     }
 
     @command('fossil.close', { repository: true})

--- a/src/commands.ts
+++ b/src/commands.ts
@@ -243,9 +243,9 @@ export class CommandCenter {
 
     @command('fossil.init')
     async init(): Promise<void> {
-        const fossilFilePath = await interaction.initFileDialog()
+        const fossilFilePath = await interaction.selectNewFossilPath();
 
-        if (fossilFilePath === undefined) {
+        if (!fossilFilePath) {
             return;
         }
         const rootPath = path.dirname(fossilFilePath) as FossilRoot;

--- a/src/commands.ts
+++ b/src/commands.ts
@@ -200,16 +200,16 @@ export class CommandCenter {
                 url = match[1] + username + ':' + userauth + '@' + match[2] as FossilURI
             }
         }
-        const parentPath = await interaction.inputFossilRootPath();
-        if (!parentPath) {
+        const rootPath = await interaction.selectFossilRootPath();
+        if (!rootPath) {
             return;
         }
 
-        const clonePromise = this.fossil.clone(url, parentPath);
+        const clonePromise = this.fossil.clone(url, rootPath);
         interaction.statusCloning(clonePromise);
 
         const repositoryPath = await clonePromise;
-        await this.askOpenRepository(repositoryPath, parentPath);
+        await this.askOpenRepository(repositoryPath, rootPath);
     }
 
     /**
@@ -258,11 +258,11 @@ export class CommandCenter {
 
     @command('fossil.open')
     async open(): Promise<void> {
-        const fossilPath = await interaction.openFileDialog()
-        if (fossilPath === undefined) {
+        const fossilPath = await interaction.selectExistingFossilPath();
+        if (!fossilPath) {
             return;
         }
-        const rootPath = await interaction.inputFossilRootPath();
+        const rootPath = await interaction.selectFossilRootPath();
         if (!rootPath) {
             return;
         }

--- a/src/commands.ts
+++ b/src/commands.ts
@@ -200,16 +200,16 @@ export class CommandCenter {
                 url = match[1] + username + ':' + userauth + '@' + match[2] as FossilURI
             }
         }
-        const rootPath = await interaction.selectFossilRootPath();
-        if (!rootPath) {
+        const fossilRoot = await interaction.selectFossilRootPath();
+        if (!fossilRoot) {
             return;
         }
 
-        const clonePromise = this.fossil.clone(url, rootPath);
+        const clonePromise = this.fossil.clone(url, fossilRoot);
         interaction.statusCloning(clonePromise);
 
-        const repositoryPath = await clonePromise;
-        await this.askOpenRepository(repositoryPath, rootPath);
+        const fossilPath = await clonePromise;
+        await this.askOpenRepository(fossilPath, fossilRoot);
     }
 
     /**
@@ -238,22 +238,22 @@ export class CommandCenter {
         const openClonedRepo = await interaction.promptOpenClonedRepo();
         if (openClonedRepo) {
             await this.openRepository(filePath, fossilRoot);
+            await this.model.tryOpenRepository(fossilRoot);
         }
     }
 
     @command('fossil.init')
     async init(): Promise<void> {
-        const fossilFilePath = await interaction.selectNewFossilPath();
+        const fossilPath = await interaction.selectNewFossilPath();
 
-        if (!fossilFilePath) {
+        if (!fossilPath) {
             return;
         }
-        const rootPath = path.dirname(fossilFilePath) as FossilRoot;
+        const fossilRoot = path.dirname(fossilPath) as FossilRoot;
 
-        // run init in the file folder in case any any artifacts appear
-        await this.fossil.init(rootPath, fossilFilePath);
-        await this.askOpenRepository(fossilFilePath, rootPath);
-        await this.model.tryOpenRepository(rootPath);
+        // run init in the file folder in case any artifacts appear
+        await this.fossil.init(fossilRoot, fossilPath);
+        await this.askOpenRepository(fossilPath, fossilRoot);
     }
 
     @command('fossil.open')

--- a/src/fossilBase.ts
+++ b/src/fossilBase.ts
@@ -8,8 +8,23 @@ import * as path from 'path';
 import * as cp from 'child_process';
 import { existsSync, appendFileSync, writeFileSync } from 'fs';
 import { groupBy, IDisposable, toDisposable, dispose, mkdirp } from "./util";
-import { EventEmitter, Event, workspace, window, Disposable, OutputChannel } from "vscode";
+import { EventEmitter, Event, workspace, window, Disposable, OutputChannel, Uri } from "vscode";
 import { interaction } from './interaction';
+
+type Distinct<T, DistinctName> = T & { __TYPE__: DistinctName };
+/** path to .fossil */
+export type FossilPath = Distinct<string, "path to .fossil">;
+/** local root */
+export type FossilRoot = Distinct<string, "local root">;
+/** something? */
+//export type FossilName = Distinct<Uri, "fossil name">;
+/** URI for the close
+ *
+ * * http[s]://[userid[:password]@]host[:port][/path]
+ * * ssh://[userid@]host[:port]/path/to/repo.fossil[?fossil=path/fossil.exe]
+ * * [file://]path/to/repo.fossil
+ */
+export type FossilURI = Distinct<string, "Fossil URI">;
 
 export interface IFossil {
     path: string;
@@ -125,12 +140,12 @@ export interface IExecutionResult {
 export async function exec(child: cp.ChildProcess, no_err_check?: boolean): Promise<IExecutionResult> {
     const disposables: IDisposable[] = [];
 
-    const once = (ee: NodeJS.EventEmitter, name: string, fn: Function) => {
+    const once = (ee: NodeJS.EventEmitter, name: string, fn: (...args: any[]) => void ) => {
         ee.once(name, fn);
         disposables.push(toDisposable(() => ee.removeListener(name, fn)));
     };
 
-    const on = (ee: NodeJS.EventEmitter, name: string, fn: Function) => {
+    const on = (ee: NodeJS.EventEmitter, name: string, fn: (...args: any[]) => void) => {
         ee.on(name, fn);
         disposables.push(toDisposable(() => ee.removeListener(name, fn)));
     };
@@ -172,8 +187,8 @@ export async function exec(child: cp.ChildProcess, no_err_check?: boolean): Prom
 export interface IFossilErrorData {
     error?: Error;
     message?: string;
-    stdout?: string;
-    stderr?: string;
+    stdout?: string; // ToDo: remove '?'
+    stderr?: string; // ToDo: remove '?'
     exitCode?: number;
     fossilErrorCode?: string;
     fossilCommand?: string;
@@ -238,6 +253,7 @@ export interface IFossilOptions {
     outputChannel: OutputChannel;
 }
 
+// ToDo: make it enum
 export const FossilErrorCodes = {
     AuthenticationFailed: 'AuthenticationFailed',
     NotAFossilRepository: 'NotAFossilRepository',
@@ -247,7 +263,8 @@ export const FossilErrorCodes = {
     BranchAlreadyExists: 'BranchAlreadyExists',
     NoUndoInformationAvailable: 'NoUndoInformationAvailable',
     UntrackedFilesDiffer: 'UntrackedFilesDiffer',
-    DefaultRepositoryNotConfigured: 'DefaultRepositoryNotConfigured'
+    DefaultRepositoryNotConfigured: 'DefaultRepositoryNotConfigured',
+    OperationMustBeforced: 'OperationMustBeforced'
 };
 
 export class Fossil {
@@ -270,29 +287,37 @@ export class Fossil {
         return this.openRepository;
     }
 
-    async init(repository: string, repoName: string): Promise<void> {
+    async init(repository: FossilRoot, repoName: FossilPath): Promise<void> {
         await this.exec(repository, ['init', repoName]);
-        return;
     }
 
-    async clone(url: string, parentPath: string): Promise<string> {
-        const folderName = url.replace(/^.*\//, '') || 'repository';
+    async clone(uri: FossilURI, parentPath: FossilRoot): Promise<FossilPath> {
+        const folderName = uri.replace(/^.*\//, '') || 'repository';
         const folderPath = path.join(parentPath, folderName + '.fossil');
 
         await mkdirp(parentPath);
-        await this.exec(parentPath, ['clone', url, folderPath]);
-        return folderPath;
+        await this.exec(parentPath, ['clone', uri, folderPath]);
+        return folderPath as FossilPath;
     }
 
-    async openClone(filePath: string, parentPath: string): Promise<void> {
-        await this.exec(parentPath, ['open', filePath]);
+    async openClone(fossilPath: FossilPath, workdir: FossilRoot): Promise<void> {
+        await this.exec(workdir, ['open', fossilPath]);
     }
 
-    async getRepositoryRoot(path: string): Promise<string> {
+    async openCloneForce(fossilPath: FossilPath, parentPath: FossilRoot): Promise<void> {
+        await this.exec(parentPath, ['open', fossilPath, '--force']);
+    }
+
+    /**
+     *
+     * @param path any path inside opened repository
+     * @returns path's root directory
+     */
+    async getRepositoryRoot(path: string): Promise<FossilRoot> {
         const result = await this.exec(path, ['stat']);
         var root = result.stdout.match(/local-root:\s*(.+)\/\s/);
-        if (root) return root[1];
-        return ""
+        if (root) return root[1] as FossilRoot;
+        return "" as FossilRoot
     }
 
     async exec(cwd: string, args: string[], options: any = {}): Promise<IExecutionResult> {
@@ -304,7 +329,8 @@ export class Fossil {
         catch (err) {
             if (err instanceof FossilError &&
                 err.fossilErrorCode !== FossilErrorCodes.NoSuchFile &&
-                err.fossilErrorCode !== FossilErrorCodes.NotAFossilRepository) {
+                err.fossilErrorCode !== FossilErrorCodes.NotAFossilRepository &&
+                err.fossilErrorCode !== FossilErrorCodes.OperationMustBeforced) {
                 const openLog = await interaction.errorPromptOpenLog(err)
                 if (openLog) {
                     this.outputChannel.show();
@@ -336,6 +362,8 @@ export class Fossil {
             }
             else if (/no such file/.test(result.stderr)) {
                 fossilErrorCode = FossilErrorCodes.NoSuchFile;
+            } else if (/--force\b/.test(result.stderr)) {
+                fossilErrorCode = FossilErrorCodes.OperationMustBeforced;
             }
 
             if (options.logErrors !== false && result.stderr) {

--- a/src/fossilBase.ts
+++ b/src/fossilBase.ts
@@ -296,7 +296,7 @@ export class Fossil {
         const folderPath = path.join(parentPath, folderName + '.fossil');
 
         await mkdirp(parentPath);
-        await this.exec(parentPath, ['clone', uri, folderPath]);
+        await this.exec(parentPath, ['clone', uri, folderPath, '--verbose']);
         return folderPath as FossilPath;
     }
 

--- a/src/fossilBase.ts
+++ b/src/fossilBase.ts
@@ -291,13 +291,11 @@ export class Fossil {
         await this.exec(repository, ['init', repoName]);
     }
 
-    async clone(uri: FossilURI, parentPath: FossilRoot): Promise<FossilPath> {
-        const folderName = uri.replace(/^.*\//, '') || 'repository';
-        const folderPath = path.join(parentPath, folderName + '.fossil');
-
-        await mkdirp(parentPath);
-        await this.exec(parentPath, ['clone', uri, folderPath, '--verbose']);
-        return folderPath as FossilPath;
+    async clone(uri: FossilURI, fossilPath: FossilPath): Promise<FossilRoot> {
+        const fossilRoot = path.dirname(fossilPath) as FossilRoot;
+        await mkdirp(fossilRoot);
+        await this.exec(fossilRoot, ['clone', uri, fossilPath, '--verbose']);
+        return fossilRoot;
     }
 
     async openClone(fossilPath: FossilPath, workdir: FossilRoot): Promise<void> {

--- a/src/interaction.ts
+++ b/src/interaction.ts
@@ -41,7 +41,7 @@ export namespace interaction {
     }
 
     /** ask user for the new .fossil file location */
-    export async function selectNewFossilPath(): Promise<FossilPath|undefined> {
+    export async function selectNewFossilPath(saveLabel: string): Promise<FossilPath|undefined> {
         const folders = workspace.workspaceFolders;
         const defaultFossilFile = Uri.file(
             folders?.length
@@ -52,7 +52,7 @@ export namespace interaction {
         const uri = await window.showSaveDialog({
             defaultUri: defaultFossilFile,
             title: 'Select New Fossil File Location',
-            saveLabel: 'Create',
+            saveLabel: saveLabel,
             filters: {
                 'All files': ['*']
             }

--- a/src/interaction.ts
+++ b/src/interaction.ts
@@ -221,6 +221,8 @@ export namespace interaction {
 
     export async function inputRepoUrl(this: void): Promise<FossilURI | undefined> {
         const url = await window.showInputBox({
+            value: 'https://fossil-scm.org/home',
+            valueSelection: [8, 100],
             prompt: localize('repourl', "Repository URI"),
             ignoreFocusOut: true
         });

--- a/src/interaction.ts
+++ b/src/interaction.ts
@@ -65,7 +65,7 @@ export namespace interaction {
      *
      * @returns fossil file uri
      */
-    export async function openFileDialog(): Promise<FossilPath|undefined> {
+    export async function selectExistingFossilPath(): Promise<FossilPath|undefined> {
         const folders = workspace.workspaceFolders;
         const defaultUri = Uri.file(
             folders?.length
@@ -260,7 +260,7 @@ export namespace interaction {
         return name;
     }
 
-    export async function inputFossilRootPath(this: void): Promise<FossilRoot | undefined> {
+    export async function selectFossilRootPath(this: void): Promise<FossilRoot | undefined> {
         const default_uri = workspace.workspaceFolders ? workspace.workspaceFolders[0].uri : undefined
         const uri = await window.showOpenDialog({
             defaultUri: default_uri,

--- a/src/interaction.ts
+++ b/src/interaction.ts
@@ -258,16 +258,18 @@ export namespace interaction {
         return name;
     }
 
-    /**
-     * poorly written function. what should it do? try to remove it!
-     */
-    export async function inputCloneParentPath(this: void): Promise<FossilRoot | undefined> {
-        var val = workspace.workspaceFolders ? workspace.workspaceFolders[0].uri.fsPath : ''
-        return await window.showInputBox({
-            prompt: localize('parent', "Parent Directory"),
-            value: val,
-            ignoreFocusOut: true
-        }) as FossilRoot;
+    export async function inputFossilRootPath(this: void): Promise<FossilRoot | undefined> {
+        const default_uri = workspace.workspaceFolders ? workspace.workspaceFolders[0].uri : undefined
+        const uri = await window.showOpenDialog({
+            defaultUri: default_uri,
+            canSelectFiles: false,
+            canSelectFolders: true,
+            canSelectMany: false,
+            title: localize('root_directory', "Select Fossil Root Directory")
+        });
+        if (uri?.length)
+            return uri[0].fsPath as FossilRoot;
+        return undefined;
     }
 
     export async function inputCloneUser(this: void): Promise<string | undefined> {

--- a/src/interaction.ts
+++ b/src/interaction.ts
@@ -8,7 +8,7 @@ import * as nls from "vscode-nls";
 import * as path from "path";
 import * as os from "os";
 import { window, QuickPickItem, workspace, ViewColumn, Uri } from "vscode";
-import { FossilUndoDetails, Path, Ref, RefType, Commit, LogEntryOptions, CommitDetails, IFileStatus } from "./fossilBase";
+import { FossilUndoDetails, Path, Ref, RefType, Commit, LogEntryOptions, CommitDetails, IFileStatus, FossilPath, FossilRoot, FossilURI } from "./fossilBase";
 import { humanise } from "./humanise";
 import { Repository, LogEntriesOptions } from "./repository";
 const localize = nls.loadMessageBundle();
@@ -30,30 +30,60 @@ export const enum WarnScenario { Merge, Update }
 export const enum CommitSources { File, Branch, Repo }
 
 export namespace interaction {
+    /**
+     *
+     * @param fsPath existing directory path
+     * @returns directory/filename.fossil
+     */
+    function suggestPath(fsPath: string): string {
+        return fsPath + '/' + (path.basename(fsPath) || 'repo_name') + '.fossil';
+    }
 
-    export async function openFileDialog(
-            fileSelect: boolean,
-            folderSelect: boolean,
-            multiSelect: boolean,
-            btnName: string
-    ): Promise<Uri[]|undefined> {
-        const homeUri = Uri.file(os.homedir());
-        const defaultUri = workspace.workspaceFolders && workspace.workspaceFolders.length > 0
-            ? Uri.file(workspace.workspaceFolders[0].uri.fsPath)
-            : homeUri;
+    /** ask user for the new .fossil file location */
+    export async function initFileDialog(): Promise<FossilPath|undefined> {
+        const folders = workspace.workspaceFolders;
+        const defaultUri = Uri.file(
+            folders?.length
+            ? suggestPath(folders[0].uri.fsPath)
+            : os.homedir()
+        );
 
-        const result = await window.showOpenDialog({
-            canSelectFiles: fileSelect,
-            canSelectFolders: folderSelect,
-            canSelectMany: multiSelect,
+        const uri = await window.showSaveDialog({
             defaultUri: defaultUri,
-            openLabel: btnName,
+            saveLabel: 'Repository Location',
             filters: {
                 'All files': ['*']
             }
         });
-        return result
+        return uri?.fsPath as FossilPath;
     }
+
+    /**
+     * ask user to open existing .fossil file
+     *
+     * @returns fossil file uri
+     */
+    export async function openFileDialog(): Promise<FossilPath|undefined> {
+        const folders = workspace.workspaceFolders;
+        const defaultUri = Uri.file(
+            folders?.length
+            ? suggestPath(folders[0].uri.fsPath)
+            : os.homedir()
+        );
+        const uri = await window.showOpenDialog({
+            defaultUri: defaultUri,
+            openLabel: 'Repository Location',
+            filters: {
+                'Fossil files': ['*.fossil'],
+                'All files': ['*']
+            },
+            canSelectMany: false
+        });
+        if (uri?.length)
+            return uri[0].fsPath as FossilPath;
+        return undefined;
+    }
+
 
     export function statusCloning(clonePromise: Promise<any>) {
         return window.setStatusBarMessage(localize('cloning', "Cloning fossil repository..."), clonePromise);
@@ -173,19 +203,26 @@ export namespace interaction {
         return choice === openOutputChannelChoice;
     }
 
-    export async function promptOpenClonedRepo(this: void) {
+    export async function promptOpenClonedRepo(this: void) : Promise<boolean> {
         const open = localize('openrepo', "Open Repository");
         const result = await window.showInformationMessage(localize('proposeopen', "Would you like to open the cloned repository?"), open);
-
         return result === open;
     }
 
-    export async function inputRepoUrl(this: void): Promise<string | undefined> {
+    export async function confirmOpenNotEmpty(this: void, dir: FossilRoot) {
+        const open = localize('openrepo', "Open Repository");
+
+        const message = localize('proposeforceopen', "The directory {0} is not empty.\nOpen repository here anyway?", dir);
+        const result = await window.showWarningMessage(message, { modal: true }, open);
+        return result === open;
+    }
+
+    export async function inputRepoUrl(this: void): Promise<FossilURI | undefined> {
         const url = await window.showInputBox({
             prompt: localize('repourl', "Repository URI"),
             ignoreFocusOut: true
         });
-        return url;
+        return url as FossilURI;
     }
 
     export async function inputPrompt(msg: string): Promise<string | undefined> {
@@ -221,13 +258,16 @@ export namespace interaction {
         return name;
     }
 
-    export async function inputCloneParentPath(this: void): Promise<string | undefined> {
+    /**
+     * poorly written function. what should it do? try to remove it!
+     */
+    export async function inputCloneParentPath(this: void): Promise<FossilRoot | undefined> {
         var val = workspace.workspaceFolders ? workspace.workspaceFolders[0].uri.fsPath : ''
         return await window.showInputBox({
             prompt: localize('parent', "Parent Directory"),
             value: val,
             ignoreFocusOut: true
-        });
+        }) as FossilRoot;
     }
 
     export async function inputCloneUser(this: void): Promise<string | undefined> {
@@ -502,7 +542,7 @@ export namespace interaction {
         return choice === respOpt;
     }
 
-    export async function handleChoices(stdout: string, limit: number): Promise<string> {
+    export async function handleChoices(stdout: string): Promise<string> {
         /* other [merge rev] changed letters.txt which local [working copy] deleted
     use (c)hanged version, leave (d)eleted, or leave (u)nresolved*/
         const [options, prompt, ..._] = stdout.split('\n').reverse();

--- a/src/interaction.ts
+++ b/src/interaction.ts
@@ -36,21 +36,23 @@ export namespace interaction {
      * @returns directory/filename.fossil
      */
     function suggestPath(fsPath: string): string {
-        return fsPath + '/' + (path.basename(fsPath) || 'repo_name') + '.fossil';
+        return path.join(fsPath,
+            (path.basename(fsPath) || 'repo_name') + '.fossil')
     }
 
     /** ask user for the new .fossil file location */
-    export async function initFileDialog(): Promise<FossilPath|undefined> {
+    export async function selectNewFossilPath(): Promise<FossilPath|undefined> {
         const folders = workspace.workspaceFolders;
-        const defaultUri = Uri.file(
+        const defaultFossilFile = Uri.file(
             folders?.length
             ? suggestPath(folders[0].uri.fsPath)
             : os.homedir()
         );
 
         const uri = await window.showSaveDialog({
-            defaultUri: defaultUri,
-            saveLabel: 'Repository Location',
+            defaultUri: defaultFossilFile,
+            title: 'Select New Fossil File Location',
+            saveLabel: 'Create',
             filters: {
                 'All files': ['*']
             }
@@ -74,7 +76,7 @@ export namespace interaction {
             defaultUri: defaultUri,
             openLabel: 'Repository Location',
             filters: {
-                'Fossil files': ['*.fossil'],
+                'Fossil Files': ['fossil'],
                 'All files': ['*']
             },
             canSelectMany: false

--- a/src/main.ts
+++ b/src/main.ts
@@ -14,7 +14,7 @@ import { FossilContentProvider } from './contentProvider';
 import * as nls from 'vscode-nls';
 import typedConfig from './config';
 
-const localize = nls.config(process.env.VSCODE_NLS_CONFIG)();
+const localize = nls.loadMessageBundle();
 
 async function init(_context: ExtensionContext, disposables: Disposable[]): Promise<void> {
     // const { name, version, aiKey } = require(context.asAbsolutePath('./package.json')) as { name: string, version: string, aiKey: string };

--- a/src/main.ts
+++ b/src/main.ts
@@ -16,7 +16,7 @@ import typedConfig from './config';
 
 const localize = nls.config(process.env.VSCODE_NLS_CONFIG)();
 
-async function init(context: ExtensionContext, disposables: Disposable[]): Promise<void> {
+async function init(_context: ExtensionContext, disposables: Disposable[]): Promise<void> {
     // const { name, version, aiKey } = require(context.asAbsolutePath('./package.json')) as { name: string, version: string, aiKey: string };
 
     const outputChannel = window.createOutputChannel('Fossil');

--- a/src/model.ts
+++ b/src/model.ts
@@ -225,6 +225,7 @@ export class Model implements Disposable {
             const repository = new Repository(this._hg.open(repositoryRoot));
 
             this.open(repository);
+            return true;
         } catch (err) {
             if (err.fossilErrorCode !== FossilErrorCodes.NotAFossilRepository) {
                 console.error('Failed to find repository:', err);

--- a/src/model.ts
+++ b/src/model.ts
@@ -131,16 +131,20 @@ export class Model implements Disposable {
 
     /**
      * Scans the first level of each workspace folder, looking
-     * for hg repositories.
+     * for fossil repositories.
      */
     private async scanWorkspaceFolders(): Promise<void> {
         for (const folder of workspace.workspaceFolders || []) {
             const root = folder.uri.fsPath;
-            const children = await new Promise<string[]>((c, e) => fs.readdir(root, (err, r) => err ? e(err) : c(r)));
+            const children = await new Promise<fs.Dirent[]>(
+                (c, e) => fs.readdir(
+                    root, {withFileTypes: true}, (err, r) => err ? e(err) : c(r)
+                )
+            );
 
             children
-                .filter(child => child !== '.hg')
-                .forEach(child => this.tryOpenRepository(path.join(root, child)));
+                .filter(child => child.isDirectory())
+                .some(child => this.tryOpenRepository(child.name));
         }
     }
 
@@ -196,10 +200,15 @@ export class Model implements Disposable {
         });
     }
 
+    /**
+     *
+     * @param path path to any file or directory in an opened repo
+     * @returns stop trying to open the repo
+     */
     @sequentialize
-    async tryOpenRepository(path: string): Promise<void> {
+    async tryOpenRepository(path: string): Promise<boolean> {
         if (this.getRepository(path)) {
-            return;
+            return true;
         }
 
         try {
@@ -210,19 +219,18 @@ export class Model implements Disposable {
             // https://github.com/Microsoft/vscode/issues/33498
 
             if (this.getRepository(repositoryRoot)) {
-                return;
+                return true;
             }
 
             const repository = new Repository(this._hg.open(repositoryRoot));
 
             this.open(repository);
         } catch (err) {
-            if (err.fossilErrorCode === FossilErrorCodes.NotAFossilRepository) {
-                return;
+            if (err.fossilErrorCode !== FossilErrorCodes.NotAFossilRepository) {
+                console.error('Failed to find repository:', err);
             }
-
-            console.error('Failed to find repository:', err);
         }
+        return false;
     }
 
     private open(repository: Repository): void {

--- a/src/repository.ts
+++ b/src/repository.ts
@@ -410,7 +410,7 @@ export class Repository implements IDisposable {
         await this.run(Operation.Status);
     }
 
-    private onFSChange(uri: Uri): void {
+    private onFSChange(_uri: Uri): void {
         if (!typedConfig.autoRefresh) {
             return;
         }

--- a/src/test/suite/index.ts
+++ b/src/test/suite/index.ts
@@ -7,7 +7,6 @@ export function run(): Promise<void> {
     const mocha = new Mocha({
         ui: 'tdd'
     });
-    mocha.useColors(true);
 
     const testsRoot = path.resolve(__dirname, '..');
 

--- a/tsconfig.json
+++ b/tsconfig.json
@@ -1,13 +1,19 @@
 {
     "compileOnSave": true,
     "compilerOptions": {
-        "target": "es6",
+        "target": "es2018",
         "lib": [
-            "es2016"
+            "es6",
+            "es2018"
         ],
         "noImplicitThis": true,
         "module": "commonjs",
         "outDir": "out",
+        //"strict": true,
+        "noImplicitReturns": true,
+        "noFallthroughCasesInSwitch": true,
+        //"noUnusedParameters": true,
+        "removeComments": true,
         "sourceMap": true,
         "strictNullChecks": true,
         "experimentalDecorators": true

